core/server/master/src/main/java/alluxio/master/file/meta/TrieNode.java

```diff
@@ -0,0 +1,263 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.master.file.meta;
+
+import alluxio.collections.Pair;
+
+import com.amazonaws.annotation.NotThreadSafe;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Stack;
+import java.util.function.BiFunction;
+
+/**
+ * TrieNode implements the Trie based on given type.
+ *
+ * It can be used in circumstances related to alluxio/ufs paths matching.
+ * @param <T> the underlying type of the TrieNode
+ */
+@NotThreadSafe
+public class TrieNode<T> {
+
+  /** mChildren stores the map from T to the child TrieNode of its children. **/
+  private final Map<T, TrieNode<T>> mChildren = new HashMap<>();
+
+  /**
+   * mIsTerminal indicates whether current TrieNode is the last node of an explicitly-inserted
+   * list of T.
+   *
+   * Here `explicitly-inserted` means the list of T is inserted by calling
+   * {@link TrieNode#insert(List, BiFunction)}. For example, T is Integer, we have a root node,
+   * and we call `root->insert(Arrays.asList(1,2,3))`. In this case, 1->2->3 is an
+   * explicitly-inserted path. So the TrieNode of `3` is the terminal node of this path.
+   *
+   * On the other hand, `implicitly-inserted` means the list of T has not been the parameter of
+   * {@link TrieNode#insert(List, BiFunction)}, whereas appearing in the Trie struct. Back to the
+   * above example, after calling `root->insert(Arrays.asList(1,2,3))`, there will be 3 path in
+   * Trie: 1, 1->2, and 1->2->3. 1 and 1->2 are implicitly-inserted, while 1->2->3 is explicitly
+   * inserted. So TrieNodes of `1` and `2` are non-terminal node.
+   */
+  private boolean mIsTerminal = false;
+
+  /**
+   * insert nodes and apply the predicate while traversing the TrieNode tree.
+   * @param nodes the nodes to be inserted
+   * @param predicate the predicate executed during traversing the existing TrieNodes. See
+   * {@link alluxio.master.file.meta.MountTable.MountTableTrieOperator} for the example of using
+   * predicate
+   * @return the last created TrieNode based on inodes
+   */
+  public TrieNode<T> insert(List<T> nodes,
+      java.util.function.BiFunction<TrieNode<T>, T, Boolean> predicate) {
```

Comment:
I think exposing this predicate in the API is confusing because people won't understand what to put there. It is also weird that we apply the predicate only when the match is not found.

What if:
1. In `TrieNode` there's only `insert(List<T> nodes)`
2. Define a `class InodeTrieNode extends TrieInode<InodeView>`
3. In `InodeTrieNode` you either define `insert(List<T> nodes, predicate)` or add a separate method like `maybeReplaceEmptyInodes(path)` after the insert.